src/gba.h

```diff
@@ -2924,6 +2925,338 @@ static uint8_t gba_audio_process_byte_write(gba_t *gba, uint32_t addr, uint8_t v
+      right_compare &= 0xFFFF;
+      left_compare_signed = (int16_t)(left_compare);
+      right_compare_signed = (int16_t)(right_compare);
+      break;
+    }
+    case 0x4:{
+      left_compare_signed = (int32_t)(left_compare);
+      right_compare_signed = (int32_t)(right_compare);
+      break;
+    }
+    case 0x6:{
+      // TODO: "Always..." codes ?
+      printf("Invalid AR if instruction\n");
+      return false;
+    }
+  }
+
+  switch(current_code&0x38){
+    case 0x8:{
+      // Equal
+      return left_compare==right_compare;
+    }
+    case 0x10:{
+      // Not equal
+      return left_compare!=right_compare;
+    }
+    case 0x18:{
+      // Signed <
+      return left_compare_signed<right_compare_signed;
+    }
+    case 0x20:{
+      // Signed >
+      return left_compare_signed>right_compare_signed;
+    }
+    case 0x28:{
+      // Unsigned <
+      return left_compare<right_compare;
+    }
+    case 0x30:{
+      // Unsigned >
+      return left_compare>right_compare;
+    }
+    case 0x38:{
+      // Logical AND
+      return left_compare&&right_compare;
+    }
+  }
+
+  return false;
+}
+bool gba_run_ar_cheat(gba_t* gba, const uint32_t* buffer, uint32_t size){
+  if(size%2!=0){
+    printf("Invalid Action Replay cheat size:%d\n",size);
+    return false;
+  }
+
+  // stack for if statements
+  // the first element is always true
+  bool if_stack[32] = {true};
+  size_t if_stack_index = 0;
+
+  for(int i=0;i<size;i+=2){
+    uint64_t code=gba_decrypt_arv3(buffer[i+1] | ((uint64_t)buffer[i] << 32));
+    uint32_t left=code>>32;
+    uint32_t right=code&0xFFFFFFFF;
+
+    if (!if_stack[if_stack_index])
+      continue;
+
+    if (right == 0x1DC0DE){
+      continue;
+    }
+
+    if (left!=0){
+      uint8_t current_code = (left>>24)&0xFF;
+      uint32_t address = ((left<<4)&0x0F000000) | (left&0x000FFFFF);
+
+      switch(current_code){
+        case 0x00:{
+          uint32_t offset=right>>8;
+          uint8_t data=right&0xFF;
+          gba_store8(gba,address+offset,data);
+          break;
+        }
+        case 0x02:{
+          uint32_t offset=right>>16;
+          uint16_t data=right&0xFFFF;
+          gba_store16(gba,address+offset*2,data);
+          break;
+        }
+        case 0x04:{
+          uint32_t data=right;
+          gba_store32(gba,address,data);
+          break;
+        }
+        case 0x40:{
+          uint32_t offset=right>>8;
+          uint8_t data=right&0xFF;
+          address=gba_read32(gba,address);
+          gba_store8(gba,address+offset,data);
+          break;
+        }
+        case 0x42:{
+          uint32_t offset=right>>16;
+          uint16_t data=right&0xFFFF;
+          address=gba_read32(gba,address);
+          gba_store16(gba,address+offset*2,data);
+          break;
+        }
+        case 0x44:{
+          uint32_t data=right;
+          address=gba_read32(gba,address);
+          gba_store32(gba,address,data);
+          break;
+        }
+        case 0x80:{
```

Comment:
Could you add a gs_button boolean variable at the top of the function and conditionals to implement the correct operation for these (and just hard code gs_button to true). That way we don't forget about the hack that we have here.

Reply:
Same for all the other GS_button cases